### PR TITLE
fix "stream not closed" problem.

### PR DESCRIPTION
**Pull request by wh81752**
While working on #456 I run into problems testing locally:

1. Downloading browsers via puppeteer takes endless amount of time (even for a single platform)
2. The command "npm run build" never finished (even if  browser download is skipped)

The first "fix" in this PR allows for rapid local testing of "npm run build" by wrapping `getBrowsers()` and `createPackage()`:

```javascript
  function isDryrun () {
    return process.env.DRY_RUN !== undefined
  }

  ..

  if (isDryrun() && true) {
    console.log('skipping browser download ..')
  } else {
    // get browser
    await getBrowsers(platforms)
  }
```

Here how to test locally:

```
$ DRY_RUN=true npm run build linux
..
skipping browser download ..
Building asciidoctor-web-pdf for: linux
..
Wrote 60.08 Mb total to linux
dry run: generated output *not* useful for production.
```

Notice the warning message that the output created in 'dry-run' mode is not particular useful as such.

The second problem has been more subtle. The documentation of https://www.npmjs.com/package/archiver states:

```
// finalize the archive (ie we are done appending files but streams have to finish yet)
// 'close', 'end' or 'finish' may be fired right after calling this method so register to them beforehand
archive.finalize();
```

So it is not enough to await the end of `archive.finalize()`, we also _need to await the end of streaming_. I introduced a new promise for that:

```
   // must not be in same dir where we are zipping
    const p0 = new Promise((resolve, reject) => {
      zipOut.on('close', function () {
        console.log(`Wrote ${Math.round(archive.pointer() / 1e4) / 1e2} Mb total to ${platform}`)
        resolve()
      })
      zipOut.on('error', reject)
      archive.on('error', reject)
    })
```
and eventually I use Promise.all() await both, the end of archive and streaming.

I tested this locally and it looks promising (correctly creates ZIp file). I wonder though how I can test my changes on my fork? Enable workflows on my fork and create somehow a test release?

### CODE REVIEW

File: tasks/prepare-binaries.js [hunk 1]
```diff
@@ -34,7 +34,7 @@ async function createPackage (platforms) {
 async function getBrowsers (platforms) {
   await Promise.all(Object.entries(platforms).map(async ([name, platform], index) => {
     const puppeteerPlatform = platform.puppeteerPlatform || name
-    await puppeteer
+    return puppeteer
```
Comment (wh81752) [inline tasks/prepare-binaries.js:37]: no need for `await` here, cause eventually all promises are awaited in by `await Promise.all()`.

File: tasks/prepare-binaries.js [hunk 2]
```diff
@@ -71,28 +71,38 @@ function copyAssets (platforms) {
...
     archive.pipe(zipOut)
+
+    // must not be in same dir where we are zipping
+    const p0 = new Promise((resolve, reject) => {
+      zipOut.on('close', function () {
+        console.log(`Wrote ${Math.round(archive.pointer() / 1e4) / 1e2} Mb total to ${platform}`)
+        resolve()
+      })
+      zipOut.on('error', reject)
+      archive.on('error', reject)
```
Comment (wh81752) [inline tasks/prepare-binaries.js:93]: listening for archive errors is strange, perhaps already included in `archive.finalize()` ??

File: tasks/prepare-binaries.js [hunk 3]
```diff
@@ -104,36 +114,60 @@ async function main (platforms) {
...
+    console.log('skipping package creation ..')
+  } else {
+    // using pkg create the binary for asciidoctor-web-pdf
+    await createPackage(platforms)
+  }
   // copy MathJax and the css/examples/fonts folders
   copyAssets(platforms)
 
-  await archive(platforms)
+  return archive(platforms)
```
Comment (wh81752) [inline tasks/prepare-binaries.js:132]: just return the promise and `await asyncMain()`.

File: tasks/prepare-binaries.js [hunk 4]
```diff
@@ -104,36 +114,60 @@ async function main (platforms) {
...
-        singleBuild[platform] = platforms[platform]
-        await main(singleBuild)
+function exit (exitcode) {
+  if (isDryrun() && exitcode === 0) {
+    console.warn('dry run: generated output *not* useful for production.')
+  }
+  process.exit(exitcode)
+}
+
+function main () {
```
Comment (Mogztter) [inline tasks/prepare-binaries.js:146]: I prefer `;(async () => { })()`, could you please revert this change?
Once we drop support for Node < 14, we will be able to use top-level await and simplify the code even more: https://v8.dev/features/top-level-await

File: tasks/prepare-binaries.js [hunk 5]
```diff
@@ -104,36 +114,60 @@ async function main (platforms) {
     fsExtra.ensureDirSync(chromiumDir)
   }
 
-  // get browser
-  await getBrowsers(platforms)
-
-  // using pkg create the binary for asciidoctor-web-pdf
-  await createPackage(platforms)
-
+  if (isDryrun()) {
```
Comment (Mogztter) [inline tasks/prepare-binaries.js:117]: If the goal is to make sure that the `archive` function works as expected then I think it would be better to write a test.

File: tasks/prepare-binaries.js [hunk 6]
```diff
@@ -71,28 +71,38 @@ function copyAssets (platforms) {
...
+      })
+      zipOut.on('error', reject)
+      archive.on('error', reject)
+    })
+
     // recursively add directory to _root_ of zip
-    archive.directory(path.join(buildDirPath, platform), false)
-    await archive.finalize()
-  }))
+    archive.directory(path.join(buildDirPath, platform), rootname, {})
```
Comment (Mogztter) [inline tasks/prepare-binaries.js:97]: We need to update the smoke test to take into account the root directory: https://github.com/Mogztter/asciidoctor-web-pdf/blob/36b4d59654742d357ff0a096355cc552e4ff2480/.github/workflows/build.yml#L74-L78

File: tasks/prepare-binaries.js (hunk 2 above)
Comment (Mogztter) [reply tasks/prepare-binaries.js:93]: from archiver documentation:

```js
// good practice to catch this error explicitly
archive.on('error', function(err) {
  throw err;
});
```

File: tasks/prepare-binaries.js [hunk 7]
```diff
@@ -71,28 +71,38 @@ function copyAssets (platforms) {
...
+    const p0 = new Promise((resolve, reject) => {
+      zipOut.on('close', function () {
+        console.log(`Wrote ${Math.round(archive.pointer() / 1e4) / 1e2} Mb total to ${platform}`)
+        resolve()
+      })
+      zipOut.on('error', reject)
+      archive.on('error', reject)
+    })
+
     // recursively add directory to _root_ of zip
```
Comment (Mogztter) [inline tasks/prepare-binaries.js:96]: we should probably update this comment

File: tasks/prepare-binaries.js [hunk 8]
```diff
@@ -71,28 +71,38 @@ function copyAssets (platforms) {
...
-    archive.directory(path.join(buildDirPath, platform), false)
-    await archive.finalize()
-  }))
+    archive.directory(path.join(buildDirPath, platform), rootname, {})
+
+    const p1 = archive.finalize()
+    return Promise.all([p0, p1])
```
Comment (Mogztter) [inline tasks/prepare-binaries.js:100]: I think we should use explicit name instead of `p0` and `p1`. What do you think about:

- `archiveFinalize`
- `zipClose`

File: tasks/prepare-binaries.js [hunk 9]
```diff
@@ -71,28 +71,38 @@ function copyAssets (platforms) {
 
 async function archive (platforms) {
   console.log('Zipping...')
-  await Promise.all(Object.keys(platforms).map(async (platform) => {
+
+  const promises = Object.keys(platforms).map(async (platform) => {
```
Comment (Mogztter) [inline tasks/prepare-binaries.js:75]: I guess we could use the following:

```js
  return Promise.all(Object.keys(platforms).map(async (platform) => { /* ... */
```

We should also probably update `getBrowsers` to be consistent:

```diff
- await Promise.all(Object.entries(platforms).map(async ([name, platform], index) => {
+ return Promise.all(Object.entries(platforms).map(async ([name, platform], index) => {
```

 .

**Comment by Mogztter** [comment]
On a side note, it seems that your email address is not associated with your GitHub account. It says "First-time contributor" and you are still missing from: https://github.com/Mogztter/asciidoctor-web-pdf/graphs/contributors

File: tasks/prepare-binaries.js (hunk 2 above)
Comment (wh81752) [reply tasks/prepare-binaries.js:93]: Yes, still strange that archive's error handling is not part of promise returned by `archive.finalize().`

File: tasks/prepare-binaries.js (hunk 5 above)
Comment (wh81752) [reply tasks/prepare-binaries.js:117]: Yes agreed, but right now I'm clueless why downloading takes endless time. It's not due to my Internet bandwith for sure. If you are unhappy with it I drop this change.

File: tasks/prepare-binaries.js (hunk 6 above)
Comment (wh81752) [reply tasks/prepare-binaries.js:97]: Good point. This change is a mistake by me. It should be in a second PR strictly about #456 while this PR is about "why the hell is `npm run build` not exiting automatically.

File: tasks/prepare-binaries.js (hunk 7 above)
Comment (wh81752) [reply tasks/prepare-binaries.js:96]: Will remove "rootname", so no need to change in this PR.

File: tasks/prepare-binaries.js (hunk 9 above)
Comment (wh81752) [reply tasks/prepare-binaries.js:75]: Changed `getBrowsers()` to return the promise. Moved the log output about downloaded browsers to `main()`.

**Comment by wh81752** [comment]
Meanwhile I figured out why `npm run build` does not exit automatically. It's because of readline's `createInterface()` as stated in https://nodejs.org/api/readline.html#readline_readline :

> Once this code is invoked, the Node.js application will not terminate until the readline.Interface is closed because the interface waits for data to be received on the input stream.

Right now this PR uses process.exit() to stop the process. Alternatively, and I guess better, is to unref `stdin` as stated:

> When creating a readline.Interface using stdin as input, the program will not terminate until it receives EOF (Ctrl+D on Linux/macOS, Ctrl+Z followed by Return on Windows). If you want your application to exit without waiting for user input, you can unref the standard input stream: `process.stdin.unref();`

This is slightly better because the exit is done gracefully, i.e. the explicit need for awaiting the end of streaming is not necessary any longer. :-(

**Comment by wh81752** [comment]
Lots of insights this morning, sorry.

> Downloading browsers via puppeteer takes endless amount of time (even for a single platform)

This because of readline not working properly in my Intellij IDEA terminal. Instead of updating the status line "Downloading browser for .." somehow my terminal is congested with "Downloading browser for linux 0% Downloading browser for linux 0 % Downloading browser for linux 1% .." and eventually 100% CPU gets used.

**Comment by Mogztter** [comment]
> This because of readline not working properly in my Intellij IDEA terminal. Instead of updating the status line "Downloading browser for .." somehow my terminal is congested with "Downloading browser for linux 0% Downloading browser for linux 0 % Downloading browser for linux 1% .." and eventually 100% CPU gets used.

I was looking for a simple way to implement a progress bar (i.e., without adding a new dependencies) but it's harder than it looks.

https://github.com/visionmedia/node-progress looks good but only supports one progress bar at a time. Maybe we should use https://www.npmjs.com/package/cli-progress or just remove the progress bar 😅

**Comment by wh81752** [comment]
Not sure whether it's worth to spend much time on the progress bar. Appears to work fine with the notable exception of IntelliJ's terminal implementation. A simple "solution" is to add a comment to prepare-binaries.js to execute in a well behaving terminal.

**Comment by wh81752** [comment]
Btw, although I spend some time on this PR I do not insist of being merged in any way. The problem of "not exiting the script" is much easier solved in PR #460.

**Comment by wh81752** [comment]
> On a side note, it seems that your email address is not associated with your GitHub account. It says "First-time contributor" and you are still missing from: https://github.com/Mogztter/asciidoctor-web-pdf/graphs/contributors

Not getting it. My email address wh81752@bitio.org has been verified by GitHub. Still not listed as commiter. How come?

**Comment by wh81752** [comment]
@Mogztter - converted this PR in a draft. Some minor things (like returning promise instead of double-awaiting) could make their way into "master" branch. Perhaps best in a new PR.

**Comment by Mogztter** [comment]
> converted this PR in a draft. Some minor things (like returning promise instead of double-awaiting) could make their way into "master" branch. Perhaps best in a new PR.

Yes, that would be great if you could isolate small changes/improvements.
I'm closing this pull request, but feel free to cherry-pick changes in a new pull request.
